packages/react-reconciler/src/ReactFiberBeginWork.js

```diff
@@ -605,6 +609,215 @@ export default function<T, P, I, TI, HI, PI, C, CC, CX, PL>(
+            // Don't scan deeper if this is a matching provider
+            nextFiber = fiber.type === workInProgress.type ? null : fiber.child;
+            break;
+          default:
+            // Traverse down.
+            nextFiber = fiber.child;
+            break;
+        }
+        if (nextFiber !== null) {
+          // Set the return pointer of the child to the work-in-progress fiber.
+          nextFiber.return = fiber;
+        } else {
+          // No child. Traverse to next sibling.
+          nextFiber = fiber;
+          while (nextFiber !== null) {
+            if (nextFiber === workInProgress) {
+              // We're back to the root of this subtree. Exit.
+              nextFiber = null;
+              break;
+            }
+            let sibling = nextFiber.sibling;
+            if (sibling !== null) {
+              nextFiber = sibling;
+              break;
+            }
+            // No more siblings. Traverse up.
+            nextFiber = nextFiber.return;
+          }
+        }
+        fiber = nextFiber;
+      }
+    }
+  }
+
+  function updateContextProvider(
+    current,
+    workInProgress,
+    renderExpirationTime,
+  ) {
+    if (enableNewContextAPI) {
+      const providerType: ReactProviderType<any> = workInProgress.type;
+      const context: ReactContext<any> = providerType.context;
+
+      const newProps = workInProgress.pendingProps;
+      const oldProps = workInProgress.memoizedProps;
+
+      if (hasLegacyContextChanged()) {
+        // Normally we can bail out on props equality but if context has changed
+        // we don't do the bailout and we have to reuse existing props instead.
+      } else if (oldProps === newProps) {
+        pushProvider(workInProgress);
+        workInProgress.stateNode = 0;
+        return bailoutOnAlreadyFinishedWork(current, workInProgress);
+      }
+      workInProgress.memoizedProps = newProps;
+
+      const newValue = newProps.value;
+
+      let changedBits: number;
+      if (oldProps === null) {
+        // Initial render
+        changedBits = MAX_SIGNED_31_BIT_INT;
+      } else {
+        const oldValue = oldProps.value;
+        // Use Object.is to compare the new context value to the old value.
+        // Inlined Object.is polyfill.
+        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/is
+        if (
+          (oldValue === newValue &&
+            (oldValue !== 0 || 1 / oldValue === 1 / newValue)) ||
+          (oldValue !== oldValue && newValue !== newValue) // eslint-disable-line no-self-compare
+        ) {
+          // No change.
+          changedBits = 0;
+        } else {
+          changedBits =
+            context.calculateChangedBits !== null
+              ? context.calculateChangedBits(oldValue, newValue)
+              : MAX_SIGNED_31_BIT_INT;
+          if (__DEV__) {
+            warning(
+              (changedBits & MAX_SIGNED_31_BIT_INT) === changedBits,
```

Comment:
Why not 32? (changedBits | 0) === changedBits. Fine that the bit field is "negative".

Reply:
IIUC, it would cause v8 to store the field in a `double`, even if you flag it as an `int` through the `|0`s.

Reply:
Yeah that’s the reason. Didn’t know this until @sebmarkbage pointed it out. https://github.com/facebook/react/pull/11818#discussion_r157120923